### PR TITLE
Add Request Header From Route

### LINKED ISSUE #2261: Using path segment in RequestHeader transform not working
### Description
Using the same configuration syntax as for QueryRouteParameter, it is not possible to set a path segment as value using the RequestHeader transform. Is this behaviour intended or is it a bug?

### To Reproduce
Configuration:
```json
  "Routes": {
    "route1": {
      "ClusterId": "cluster1",
      "Match": {
        "Path": "/api/{plugin}/{**remainder}"
      },
      "Transforms": [
        {
          "QueryRouteParameter": "q",
          "Append": "plugin"
        },
        {
          "RequestHeader": "header1",
          "Set": "plugin"
        }
      ]
    }
  }
```
`GET <domain>/api/foo/bar` returns
```json
{
    "method": "GET",
    "path": "/api/foo/bar",
    "query": {
        "q": "foo"
    },
    "headers": {
        "header1": "plugin"
    }
}
```

### Expected Behaviour
`GET <domain>/api/foo/bar` returns
```json
{
    "method": "GET",
    "path": "/api/foo/bar",
    "query": {
        "q": "foo"
    },
    "headers": {
        "header1": "foo"
    }
}
```

### PR DESCRIPTION
**Pull request by daniel-steinmann**
Resolves #2261
Adding a transform to allow using route values in headers to achieve the same functionality as with QueryRouteParameter.

### CODE REVIEW

File: src/ReverseProxy/Transforms/RequestHeaderTransform.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Primitives;
+
+namespace Yarp.ReverseProxy.Transforms;
+
+public enum RequestHeaderTransformMode
```
Comment (Tratcher) [inline src/ReverseProxy/Transforms/RequestHeaderTransform.cs:7]: Elsewhere we just use a bool:
https://github.com/microsoft/reverse-proxy/blob/debbe0928c5a85b6ca6e6378a2d6458df5550ff8/src/ReverseProxy/Transforms/RequestHeaderValueTransform.cs#L15

File: docs/docfx/articles/transforms.md [hunk 2]
```diff
@@ -459,6 +459,44 @@ MyHeader: MyValue
 This sets or appends the value for the named header. Set replaces any existing header. Append adds an additional header with the given value.
 Note: setting "" as a header value is not recommended and can cause an undefined behavior.
 
+### RequestHeaderFromRoute
```
Comment (Tratcher) [inline docs/docfx/articles/transforms.md:462]: This works like PathRouteValuesTransform, should they use the same naming scheme? `RequestHeaderRouteValue`.

File: src/ReverseProxy/Transforms/RequestHeaderTransform.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+{
+    Append,
+    Set
+}
+
+public abstract class RequestHeaderTransform : RequestTransform
```
Comment (Tratcher) [inline src/ReverseProxy/Transforms/RequestHeaderTransform.cs:13]: Can RequestHeaderValueTransform derive from this?

Comment (daniel-steinmann) [reply src/ReverseProxy/Transforms/RequestHeaderTransform.cs:7]: RequestHeaderValueTransform inherits from RequestHeaderTransform now.

File: docs/docfx/articles/transforms.md (hunk 2 above)
Comment (daniel-steinmann) [reply docs/docfx/articles/transforms.md]: I agree, changed it 👍

File: src/ReverseProxy/Transforms/RequestHeaderTransform.cs (hunk 1 above)
Comment (daniel-steinmann) [reply src/ReverseProxy/Transforms/RequestHeaderTransform.cs]: Replaced RequestHeaderTransformMode with bool.

**Comment by daniel-steinmann** [comment]
> @daniel-steinmann please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree

**Review by Tratcher — APPROVED** [review]

File: src/ReverseProxy/Transforms/RequestHeaderValueTransform.cs [hunk 4]
```diff
@@ -10,26 +10,20 @@ namespace Yarp.ReverseProxy.Transforms;
...
     }
 
-    internal string HeaderName { get; }
-
     internal string Value { get; }
 
-    internal bool Append { get; }
-
     /// <inheritdoc/>
     public override ValueTask ApplyAsync(RequestTransformContext context)
```
[the remark this answers is not in the document]
Comment (daniel-steinmann) [reply src/ReverseProxy/Transforms/RequestHeaderValueTransform.cs:27]: Right, the override was redundant. Do you have a particular case in mind where removing the override altogether would be problematic? For now I left it in with a call to base as you suggested.

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/ReverseProxy/Transforms/RequestHeaderValueTransform.cs:27]: Just being cautious about binary breaking changes. A library could be compiled against the existing overload and break when it's removed.

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks

**Pull request merged**